server/player/provider/inventory.go

```diff
@@ -0,0 +1,173 @@
+package provider
+
+import (
+	"github.com/df-mc/dragonfly/server/item"
+	"github.com/df-mc/dragonfly/server/player"
+	"github.com/df-mc/dragonfly/server/world"
+)
+
+func invToData(data player.InventoryData) jsonInventoryData {
+	d := jsonInventoryData{
+		Mainhand: data.Mainhand,
+		Offhand:  itemToData(data.Offhand),
+	}
+	for slot, i := range data.Items {
+		itemData := itemToData(i)
+		if itemData == nil {
+			continue
+		}
+		itemData["slot"] = slot
+		d.Items = append(d.Items, itemData)
+	}
+	for slot, i := range data.Armour {
+		itemData := itemToData(i)
+		if itemData == nil {
+			d.Armour[slot] = nil
+		}
+		d.Armour[slot] = itemData
+	}
+	return d
+}
+
+func dataToInv(data jsonInventoryData) player.InventoryData {
+	d := player.InventoryData{
+		Mainhand: data.Mainhand,
+		Offhand:  dataToItem(data.Offhand),
+		Items:    make([]item.Stack, 36),
+		Armour:   [4]item.Stack{},
+	}
+	for _, i := range data.Items {
+		slot, ok := readInt("slot", i)
+		if !ok {
+			continue
+		}
+		d.Items[slot] = dataToItem(i)
+	}
+	for slot, i := range data.Armour {
+		d.Armour[slot] = dataToItem(i)
+	}
+	return d
+}
+
+func itemToData(stack item.Stack) map[string]interface{} {
+	data := make(map[string]interface{})
+	if stack.Empty() {
+		return nil
+	}
+
+	if b, ok := stack.Item().(world.Block); ok {
+		data["block"], data["block_properties"] = b.EncodeBlock()
+	} else {
+		data["item"], data["item_meta"] = stack.Item().EncodeItem()
+	}
+
+	data["count"] = stack.Count()
+	if len(stack.Values()) > 0 {
+		data["nbt"] = stack.Values()
```

Comment:
Could this key be "dragonflyData" instead of "nbt"? Would be more consistent with nbtconv.